rust-install/src/dist.rs

```diff
@@ -268,3 +273,84 @@ pub fn download_hash(url: &str, cfg: DownloadCfg) -> Result<String> {
+// an upgrade then all the existing components will be upgraded.
+//
+// Returns the manifest's hash if anything changed.
+pub fn update_from_dist<'a>(download: DownloadCfg<'a>,
+                            update_hash: Option<&Path>,
+                            toolchain: &str,
+                            prefix: &InstallPrefix,
+                            add: &[Component],
+                            remove: &[Component],
+                            ) -> Result<Option<String>> {
+
+    let ref toolchain = try!(ToolchainDesc::from_str(toolchain).ok_or(Error::InvalidToolchainName));
+    let trip = try!(toolchain.target_triple().ok_or_else(|| Error::UnsupportedHost(toolchain.full_spec())));
+
+    let manifestation = try!(Manifestation::open(prefix.clone(), &trip));
+
+    let changes = Changes {
+        add_extensions: add.to_owned(),
+        remove_extensions: remove.to_owned(),
+    };
+
+    match dl_v2_manifest(download, update_hash, toolchain) {
+        Ok(Some((m, hash))) => {
+            match try!(manifestation.update(&m, changes, &download.temp_cfg, download.notify_handler.clone())) {
+                UpdateStatus::Unchanged => Ok(None),
+                UpdateStatus::Changed => Ok(Some(hash)),
+            }
+        }
+        Ok(None) => Ok(None),
+        Err(Error::Utils(utils::Error::DownloadingFile {
+            error: utils::raw::DownloadError::Status(hyper::status::StatusCode::NotFound),
+            ..
+        })) => {
+            // If the v2 manifest is not found then try v1
+            let manifest = try!(dl_v1_manifest(download,toolchain));
+            match try!(manifestation.update_v1(&manifest, update_hash,
+                                               &download.temp_cfg, download.notify_handler.clone())) {
+                None => Ok(None),
+                Some(hash) => Ok(Some(hash)),
+            }
+        }
+        Err(e) => Err(e)
+    }
+}
+
+fn dl_v2_manifest<'a>(download: DownloadCfg<'a>,
+                      update_hash: Option<&Path>,
+                      toolchain: &ToolchainDesc) -> Result<Option<(ManifestV2, String)>> {
+    let ref manifest_url = if let Some(ref d) = toolchain.date {
+        format!("{}/{}/channel-rust-{}.toml", download.dist_root, d, toolchain.channel)
+    } else {
+        format!("{}/channel-rust-{}.toml", download.dist_root, toolchain.channel)
+    };
+    let manifest_dl = try!(download_and_check(&manifest_url,
+                                              update_hash, ".toml", download));
+    let (manifest_file, manifest_hash) = if let Some(m) = manifest_dl { m } else { return Ok(None) };
+    let manifest_str = try!(utils::read_file("manifest", &manifest_file));
+    let manifest = try!(ManifestV2::parse(&manifest_str));
+
+    Ok(Some((manifest, manifest_hash)))
+}
+
+fn dl_v1_manifest<'a>(download: DownloadCfg<'a>,
+                      toolchain: &ToolchainDesc) -> Result<Vec<String>> {
+    let root_url = if let Some(ref d) = toolchain.date {
```

Comment:
The existing method on `ToolchainDesc` does somewhat less than this function. I also don't think the toolchain descriptor should have the responsibility to download things from the internet, but should be a simple data type. Do you mind if I just delete the original method?